### PR TITLE
Make sure the new datetime functions work with literal

### PR DESCRIPTION
**Pull request by qnkhuat**
datetime-add, datetime-subtract, temporal-extract, convert-timezone should works with literal date

### CODE REVIEW

File: src/metabase/query_processor/middleware/wrap_value_literals.clj [hunk 1]
```diff
@@ -123,14 +129,16 @@
...
+    [:datetime-diff (parse-date-if-string x) (parse-date-if-string y) unit]
+
+    [:temporal-extract field & args]
+    (into [:temporal-extract (parse-date-if-string field)] args)
+
+    [(clause :guard #{:datetime-add :datetime-subtract}) field amount unit]
+    [clause (parse-date-if-string field) amount unit]
+
+    [:convert-timezone field & args]
+    (into [:convert-timezone (parse-date-if-string field)] args)
```
Comment (calherries) [inline src/metabase/query_processor/middleware/wrap_value_literals.clj:141]: These could all be combined, right?
```suggestion
    [:temporal-extract field & args]
    (into [:temporal-extract (parse-date-if-string field)] args)

    [(clause :guard #{:datetime-add :datetime-subtract}) field amount unit]
    [clause (parse-date-if-string field) amount unit]

    [(clause :guard #{:datetime-add :datetime-subtract :convert-timezone :temporal-extract}) (field :guard string?) & args]
    (into [clause (add-type-info (u.date/parse y) nil)] args)
```

Comment (qnkhuat) [reply src/metabase/query_processor/middleware/wrap_value_literals.clj:141]: clean!

**Review by metamben — APPROVED** [review]
Looks like a strict improvement.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj [hunk 2]
```diff
@@ -120,7 +119,28 @@
                 {:keys [expected-fn query-fn]}
                 extraction-test-cases]
           (testing (format "extract %s function works as expected on %s column for driver %s" op col-type driver/*driver*)
-            (is (= (set (expected-fn op)) (set (test-temporal-extract (query-fn op field-id)))))))))))
+            (is (= (set (expected-fn op)) (set (test-temporal-extract (query-fn op field-id)))))))))
+
+    ;; not really sure why vertica is failing...
```
Comment (calherries) [inline test/metabase/query_processor_test/date_time_zone_functions_test.clj:124]: What's the error with vertica?

Comment (qnkhuat) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:124]:
```
[Vertica][VJDBC](3459) ERROR: Function catalog.date_part(unknown, unknown) is not unique
```

even though the query has no `date_part` fn

```
SELECT "public"."times_mixed_times"."id" AS "id",
          "public"."times_mixed_times"."index" AS "index",
          "public"."times_mixed_times"."dt" AS "dt",
          "public"."times_mixed_times"."dt_tz" AS "dt_tz",
          "public"."times_mixed_times"."d" AS "d",
          CAST("public"."times_mixed_times"."as_dt" AS timestamp) AS "as_dt",
          CAST("public"."times_mixed_times"."as_d" AS date) AS "as_d",
          year(?) AS "get-year",
          CAST(extract(quarter
                       from ?) AS integer) AS "get-quarter",
          CAST(extract(month
                       from ?) AS integer) AS "get-month",
          CAST(extract(day
                       from ?) AS integer) AS "get-day",
          CASE
              WHEN ((dayofweek_iso(?) + 1) % 7) = 0 THEN 7
              ELSE ((dayofweek_iso(?) + 1) % 7)
          END AS "get-day-of-week",
          CAST(extract(hour
                       from ?) AS integer) AS "get-hour",
          CAST(extract(minute
                       from ?) AS integer) AS "get-minute",
          second(?) AS "get-second"
   FROM "public"."times_mixed_times"
```

Comment (calherries) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj:124]: My guess is `extract` is syntactic sugar for `date_part`, and the error comes from not casting the timestamps correctly. Here's a minimal reproduction of the error:

```
(ffirst (:rows (dev/query-jdbc-db :vertica "select extract(epoch from '2022-10-01T01:30:00')")))
;; => Execution error (SyntaxErrorException) at com.vertica.util.ServerErrorData/buildException (REPL:-1).
;;    [Vertica][VJDBC](3459) ERROR: Function catalog.date_part(unknown, unknown) is not unique

(ffirst (:rows (dev/query-jdbc-db :vertica "select extract(epoch from '2022-10-01T01:30:00'::timestamp)")))
;; => 1664587800.000000M
```

File: modules/drivers/vertica/src/metabase/driver/vertica.clj [hunk 3]
```diff
@@ -75,7 +75,7 @@
     expr))
 
 (defn- date-trunc [unit expr] (hsql/call :date_trunc (hx/literal unit) (cast-timestamp expr)))
-(defn- extract    [unit expr] (hsql/call :extract    unit              expr))
+(defn- extract    [unit expr] (hsql/call :extract    unit              (cast-timestamp expr)))
```
Comment (calherries) [inline modules/drivers/vertica/src/metabase/driver/vertica.clj:78]: Having to change `extract` to get vertica to work here suggests to me that datetime literals aren't being spliced into the SQL query correctly for vertica.

Does vertica need implementations for `metabase.driver.sql.util.unprepare/unprepare-value`? e.g.
https://github.com/metabase/metabase/blob/b2791cf8d5d8022ac9d4499332c5ea2f5cceb3be/modules/drivers/snowflake/src/metabase/driver/snowflake.clj#L346

Comment (qnkhuat) [reply modules/drivers/vertica/src/metabase/driver/vertica.clj:78]: unprepare is actually irrelevant here.

unprepare is used for the `convert to SQL` feature where we convert a native with params, to native without params.

File: test/metabase/query_processor_test/date_time_zone_functions_test.clj (hunk 2 above)
Comment (qnkhuat) [reply test/metabase/query_processor_test/date_time_zone_functions_test.clj]: yep!.  fixed in https://github.com/metabase/metabase/pull/26706/commits/b2aeb6b237bdf12f4ca7d7c955d5be3af9b183d2

File: modules/drivers/vertica/src/metabase/driver/vertica.clj (hunk 3 above)
Comment (calherries) [reply modules/drivers/vertica/src/metabase/driver/vertica.clj:78]: Right. It still seems strange that vertica's datetime literals aren't being spliced with the same effect as `unprepare-value`. But I suppose it's an improvement given we're already doing it for `date-trunc`.

**Review by calherries — APPROVED** [review]

File: test/metabase/util/honeysql_extensions_test.clj [hunk 4]
```diff
@@ -192,6 +192,7 @@
     typed-form     "date"   false
     typed-form     nil      false
     typed-form     #"tex.*" true
+    nil            #"tex.*" false
     typed-form     #"int*"  false
```
Comment (calherries) [inline test/metabase/util/honeysql_extensions_test.clj:196]: If database-type is nilable, there should be a test for this here

Comment (qnkhuat) [reply test/metabase/util/honeysql_extensions_test.clj:196]: isn't the test on line 193 is testing with `nil` database-type?

or you mean `(is-of-type? nil nil)`?

Comment (calherries) [reply test/metabase/util/honeysql_extensions_test.clj:196]: Ah, duh. Never mind

Comment (qnkhuat) [reply test/metabase/util/honeysql_extensions_test.clj:196]: either way I'll add the `nil nil` case to the test.

File: src/metabase/util/honeysql_extensions.clj [hunk 5]
```diff
@@ -224,7 +224,7 @@
   [honeysql-form database-type]
   (let [form-type (some-> honeysql-form type-info type-info->db-type str/lower-case)]
     (if (instance? java.util.regex.Pattern database-type)
-      (some? (re-find database-type form-type))
+      (and (some? form-type) (some? (re-find database-type form-type)))
       (= form-type
          (some-> database-type name str/lower-case)))))
```
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/util/honeysql_extensions.clj:229]: Never mind, I see that database-type is not nillable in the true branch now.

File: test/metabase/util/honeysql_extensions_test.clj (hunk 4 above)
Comment (calherries) [reply test/metabase/util/honeysql_extensions_test.clj:196]: Perhaps `is-of-type?` could be refactored to make the result of `nil` database-type more obvious. e.g. turning the `if` into a `cond`

**Pull request merged**